README.md
# drdb9\_exporter

This is a Prometheus exporter that, when running on a node, checks the status
of that node's DRBD volumes.

## Installation

To build the deb file:

```
debuild -b -uc -us
```

To build the binary:

```bash
make
```

```bash
CGO_ENABLED=0 go build
```

To build the Docker container:

```bash
CGO_ENABLED=0 make GOLDFLAGS="-w -linkmode external -extldflags -static" && docker build .
```

## Usage

Command-line options:

```plain
Usage of drbd9_exporter:
  -web.listen-address string
    	Address on which to expose metrics and web interface (default ":9481")
  -web.telemetry-path string
    	Path under which to expose metrics. (default "/metrics")
```

## Assumptions

This exporter assummes that you are running DRBD 9.0 and have the 
debugfs mounted, most distros do, but if you don't then you can 
manually mount the debugfs by doing:

```bash
mount -t debugfs nodev /sys/kernel/debug
```

## Example export

```plain
# HELP drbd_activitylog_writes Number of updates of the activity log area of the meta data.
# TYPE drbd_activitylog_writes gauge
drbd_activitylog_writes{connection="n2",id="0",resource="r0"} 0
drbd_activitylog_writes{connection="n3",id="0",resource="r0"} 0
# HELP drbd_bitmap_writes Number of updates of the bitmap area of the meta data.
# TYPE drbd_bitmap_writes gauge
drbd_bitmap_writes{connection="n2",id="0",resource="r0"} 0
drbd_bitmap_writes{connection="n3",id="0",resource="r0"} 0
# HELP drbd_connection_established a
# TYPE drbd_connection_established gauge
drbd_connection_established{connection="n2",id="0",resource="r0"} 1
drbd_connection_established{connection="n3",id="0",resource="r0"} 1
# HELP drbd_connection_unhealthy a
# TYPE drbd_connection_unhealthy gauge
drbd_connection_unhealthy{connection="n2",id="0",resource="r0"} 0
drbd_connection_unhealthy{connection="n3",id="0",resource="r0"} 0
# HELP drbd_disk_read_bytes Net data read from local hard disk.
# TYPE drbd_disk_read_bytes gauge
drbd_disk_read_bytes{connection="n2",id="0",resource="r0"} 262056
drbd_disk_read_bytes{connection="n3",id="0",resource="r0"} 262056
# HELP drbd_disk_written_bytes Net data written on local hard disk.
# TYPE drbd_disk_written_bytes gauge
drbd_disk_written_bytes{connection="n2",id="0",resource="r0"} 0
drbd_disk_written_bytes{connection="n3",id="0",resource="r0"} 0
# HELP drbd_epochs Number of Epochs currently on the fly.
# TYPE drbd_epochs gauge
drbd_epochs{connection="n2",id="0",resource="r0"} 1
drbd_epochs{connection="n3",id="0",resource="r0"} 1
# HELP drbd_local_disk_up_to_date a
# TYPE drbd_local_disk_up_to_date gauge
drbd_local_disk_up_to_date{connection="n2",id="0",resource="r0"} 1
drbd_local_disk_up_to_date{connection="n3",id="0",resource="r0"} 1
# HELP drbd_local_pending Number of open requests to the local I/O sub-system.
# TYPE drbd_local_pending gauge
drbd_local_pending{connection="n2",id="0",resource="r0"} 0
drbd_local_pending{connection="n3",id="0",resource="r0"} 0
# HELP drbd_network_received_bytes Volume of net data received by the partner via the network connection.
# TYPE drbd_network_received_bytes gauge
drbd_network_received_bytes{connection="n2",id="0",resource="r0"} 0
drbd_network_received_bytes{connection="n3",id="0",resource="r0"} 0
# HELP drbd_network_sent_bytes Volume of net data sent to the partner via the network connection.
# TYPE drbd_network_sent_bytes gauge
drbd_network_sent_bytes{connection="n2",id="0",resource="r0"} 131028
drbd_network_sent_bytes{connection="n3",id="0",resource="r0"} 131028
# HELP drbd_out_of_sync_bytes Amount of data known to be out of sync.
# TYPE drbd_out_of_sync_bytes gauge
drbd_out_of_sync_bytes{connection="n2",id="0",resource="r0"} 0
drbd_out_of_sync_bytes{connection="n3",id="0",resource="r0"} 0
# HELP drbd_remote_unacknowledged Number of requests received by the partner via the network connection, but that have not yet been answered.
# TYPE drbd_remote_unacknowledged gauge
drbd_remote_unacknowledged{connection="n2",id="0",resource="r0"} 0
drbd_remote_unacknowledged{connection="n3",id="0",resource="r0"} 0
```

debian/changelog
drbd9-exporter (0.1) unstable; urgency=medium

  * Initial release.

 -- Ben Cartwright-Cox <bencartwrightcox@monzo.com>  Tue, 19 Nov 2019 16:01:29 +0100

drbd_test.go
package main

import "testing"

var (
	testCase1 = []byte(` 1: cs:Established ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
	ns:0 nr:119598720 dw:161539164 dr:0 al:2125 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
	resync: used:0/61 hits:25476 misses:922 starving:0 locked:0 changed:592
	act_log: used:0/1237 hits:13283342 misses:15109 starving:0 locked:0 changed:12057
	blocked on activity log: 0/0/0`)
	testCase2 = []byte(` 0: cs:Established ro:Secondary/Secondary ds:UpToDate/UpToDate C r-----
	ns:0 nr:0 dw:158780236 dr:0 al:1950 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
	resync: used:0/61 hits:2 misses:285 starving:0 locked:0 changed:285
	act_log: used:0/1237 hits:13371654 misses:15913 starving:0 locked:0 changed:11885
	blocked on activity log: 0/0/0`)
	testCase3 = []byte(` 1: cs:Established ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
	ns:119876764 nr:0 dw:154278296 dr:14701504 al:2070 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
	resync: used:0/61 hits:25566 misses:719 starving:0 locked:0 changed:485
	act_log: used:0/1237 hits:13318630 misses:41690 starving:0 locked:0 changed:12063
	blocked on activity log: 0/0/0`)
	testCase4 = []byte(` 1: cs:SyncSource ro:Primary/Secondary ds:UpToDate/Inconsistent C s-----
	ns:119876764 nr:0 dw:154278296 dr:14701504 al:2070 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
	resync: used:0/61 hits:25566 misses:719 starving:0 locked:0 changed:485
	act_log: used:0/1237 hits:13318630 misses:41690 starving:0 locked:0 changed:12063
	blocked on activity log: 0/0/0`)
)

func TestDrbdParse1(t *testing.T) {
	output := drbdConnection{}
	err := parseProcDRBD(testCase1, &output)
	if err != nil {
		t.Fatalf("failed to parse legit output %s", err.Error())
	}

	if output.ConnectionStatus != "Established" {
		t.Fatalf("mis-parsed testCase1 Expected 'Established' got %s", output.ConnectionStatus)
	}
	if output.MyDiskStatus != "UpToDate" {
		t.Fatalf("mis-parsed testCase1 Expected 'UpToDate' got %s", output.MyDiskStatus)
	}
	if output.RemoteRole != "Primary" {
		t.Fatalf("mis-parsed testCase1 Expected 'Primary' got %s", output.RemoteRole)
	}
}

func TestDrbdParse2(t *testing.T) {
	output := drbdConnection{}
	err := parseProcDRBD(testCase2, &output)
	if err != nil {
		t.Fatalf("failed to parse legit output %s", err.Error())
	}

	if output.ConnectionStatus != "Established" {
		t.Fatalf("mis-parsed testCase2 Expected 'Established' got %s", output.ConnectionStatus)
	}
	if output.MyDiskStatus != "UpToDate" {
		t.Fatalf("mis-parsed testCase2 Expected 'UpToDate' got %s", output.MyDiskStatus)
	}
	if output.RemoteRole != "Secondary" {
		t.Fatalf("mis-parsed testCase2 Expected 'Secondary' got %s", output.RemoteRole)
	}

	foundAl := false
	for _, v := range output.KVs {
		if v.Name == "al" {
			foundAl = true
			if v.Value != 1950 {
				t.Fatalf("Did not parse value correctly from KVs %#v", v)
			}
		}
	}
	if !foundAl {
		t.Fatalf("Expected AL kv to have been parsed, didnt exist: %#v", output.KVs)
	}
}

func TestDrbdParse3(t *testing.T) {
	output := drbdConnection{}
	err := parseProcDRBD(testCase3, &output)
	if err != nil {
		t.Fatalf("failed to parse legit output %s", err.Error())
	}

	if output.ConnectionStatus != "Established" {
		t.Fatalf("mis-parsed testCase3 Expected 'Established' got %s", output.ConnectionStatus)
	}
	if output.MyDiskStatus != "UpToDate" {
		t.Fatalf("mis-parsed testCase3 Expected 'UpToDate' got %s", output.MyDiskStatus)
	}
	if output.RemoteRole != "Secondary" {
		t.Fatalf("mis-parsed testCase3 Expected 'Secondary' got %s", output.RemoteRole)
	}
	if output.MyRole != "Primary" {
		t.Fatalf("mis-parsed testCase3 Expected 'Primary' got %s", output.RemoteRole)
	}
}

func TestDrbdParse4(t *testing.T) {
	output := drbdConnection{}
	err := parseProcDRBD(testCase4, &output)
	if err != nil {
		t.Fatalf("failed to parse legit output %s", err.Error())
	}

	if output.ConnectionStatus != "SyncSource" {
		t.Fatalf("mis-parsed testCase4 Expected 'SyncSource' got %s", output.ConnectionStatus)
	}
	if output.MyDiskStatus != "UpToDate" {
		t.Fatalf("mis-parsed testCase4 Expected 'UpToDate' got %s", output.MyDiskStatus)
	}
	if output.RemoteRole != "Secondary" {
		t.Fatalf("mis-parsed testCase4 Expected 'Secondary' got %s", output.RemoteRole)
	}
	if output.MyRole != "Primary" {
		t.Fatalf("mis-parsed testCase4 Expected 'Primary' got %s", output.RemoteRole)
	}

	if !output.Suspended {
		t.Fatal("mis-parsed testCase4 Expected to be suspended, was not")
	}
}


collector.go
package main

import (
	"fmt"
	"log"

	"github.com/prometheus/client_golang/prometheus"
)

var (
	connectionEstablished = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "drbd_connection_established",
			Help: "a",
		},
		[]string{"resource", "connection", "id"},
	)
	diskUpToDate = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "drbd_local_disk_up_to_date",
			Help: "a",
		},
		[]string{"resource", "connection", "id"},
	)
	connectionUnhealthy = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "drbd_connection_unhealthy",
			Help: "a",
		},
		[]string{"resource", "connection", "id"},
	)
	gauges = map[string]*prometheus.GaugeVec{
		"ns": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_network_sent_bytes",
				Help: "Volume of net data sent to the partner via the network connection.",
			},
			[]string{"resource", "connection", "id"}),
		"nr": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_network_received_bytes",
				Help: "Volume of net data received by the partner via the network connection.",
			},
			[]string{"resource", "connection", "id"}),
		"dw": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_disk_written_bytes",
				Help: "Net data written on local hard disk.",
			},
			[]string{"resource", "connection", "id"}),
		"dr": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_disk_read_bytes",
				Help: "Net data read from local hard disk.",
			},
			[]string{"resource", "connection", "id"}),
		"al": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_activitylog_writes",
				Help: "Number of updates of the activity log area of the meta data.",
			},
			[]string{"resource", "connection", "id"}),
		"bm": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_bitmap_writes",
				Help: "Number of updates of the bitmap area of the meta data.",
			},
			[]string{"resource", "connection", "id"}),

		"lo": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_local_pending",
				Help: "Number of open requests to the local I/O sub-system.",
			},
			[]string{"resource", "connection", "id"}),
		"pe": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_remote_pending",
				Help: "Number of requests sent to the partner, but that have not yet been answered by the latter.",
			},
			[]string{"resource", "connection", "id"}),
		"ua": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_remote_unacknowledged",
				Help: "Number of requests received by the partner via the network connection, but that have not yet been answered.",
			},
			[]string{"resource", "connection", "id"}),
		"ap": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_application_pending",
				Help: "Number of block I/O requests forwarded to DRBD, but not yet answered by DRBD.",
			},
			[]string{"resource", "connection", "id"}),
		"ep": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_epochs",
				Help: "Number of Epochs currently on the fly.",
			},
			[]string{"resource", "connection", "id"}),
		"oos": prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "drbd_out_of_sync_bytes",
				Help: "Amount of data known to be out of sync.",
			},
			[]string{"resource", "connection", "id"}),
	}
)

//Collector implements the prometheus.Collector interface.
type Collector struct{}

//Describe implements the prometheus.Collector interface.
func (c Collector) Describe(ch chan<- *prometheus.Desc) {
	connectionEstablished.Describe(ch)
	diskUpToDate.Describe(ch)
	connectionUnhealthy.Describe(ch)
	for _, v := range gauges {
		v.Describe(ch)
	}
}

//Collect implements the prometheus.Collector interface.
func (c Collector) Collect(ch chan<- prometheus.Metric) {
	err := c.measure()
	//only report data when measurement was successful
	if err == nil {
		connectionEstablished.Collect(ch)
		diskUpToDate.Collect(ch)
		connectionUnhealthy.Collect(ch)

		for _, v := range gauges {
			v.Collect(ch)
		}
	} else {
		log.Println("ERROR:", err)
		return
	}
}

var errNoStatus = fmt.Errorf("No DRBD volumes detected")

func (c Collector) measure() error {
	connections := getAllDRDBstatues()
	if len(connections) == 0 {
		return errNoStatus
	}

	for _, connection := range connections {

		isEstablished := (connection.ConnectionStatus == "Established" || connection.ConnectionStatus == "SyncSource" || connection.ConnectionStatus == "SyncTarget")
		if isEstablished {
			connectionEstablished.WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(1)
			connectionUnhealthy.WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(0)
		} else {
			connectionEstablished.WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(0)
			connectionUnhealthy.WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(1)
		}

		if connection.MyDiskStatus != "UpToDate" {
			diskUpToDate.WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(0)
		} else {
			diskUpToDate.WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(1)
		}

		for _, kv := range connection.KVs {
			if gauges[kv.Name] != nil {
				gauges[kv.Name].WithLabelValues(connection.Resource, connection.RemoteHost, connection.ResourceID).Set(kv.Value)
			}
		}

	}
	return nil
}


drbd9_exporter.go
package main

import (
	"flag"
	"log"
	"net/http"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/version"
)

var (
	listenAddress = flag.String("web.listen-address", ":9481", "Address on which to expose metrics and web interface")
	metricsPath   = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
)

func main() {
	flag.Parse()

	log.Print("Starting node_exporter", version.Info())
	log.Print("Build context", version.BuildContext())
	prometheus.MustRegister(Collector{})
	handler := promhttp.HandlerFor(prometheus.DefaultGatherer,
		promhttp.HandlerOpts{})

	http.Handle(*metricsPath, handler)

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
			<head><title>DRBD9 Exporter</title></head>
			<body>
			<h1>DRBD9 Exporter</h1>
			<p><a href="` + *metricsPath + `">Metrics</a></p>
			</body>
			</html>`))
	})

	log.Print("Listening on", *listenAddress)
	if err := http.ListenAndServe(*listenAddress, nil); err != nil {
		log.Fatal(err)
	}
}


drbd.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

type drbdConnection struct {
	// Infomation about the connection
	Resource   string
	RemoteHost string
	ResourceID string
	// Properties / Status of the connection
	ConnectionStatus string
	MyRole           string
	RemoteRole       string
	MyDiskStatus     string
	RemoteDiskStatus string
	Suspended        bool
	KVs              []drbdConnectionKV
}

type drbdConnectionKV struct {
	Name  string
	Value float64
}

func getAllDRDBstatues() []drbdConnection {
	connections := make([]drbdConnection, 0)

	filepath.Walk("/sys/kernel/debug/drbd/resources/", func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if info.Name() != "proc_drbd" {
			return nil
		}

		// /sys/kernel/debug/drbd/resources/$Resource/connections/$RemoteHost/$ResourceID/proc_drbd
		//
		// []string{"", "sys", "kernel", "debug", "drbd", "resources", "$Resource", "connections", "$RemoteHost", "$ResourceID", "proc_drbd"}
		pathSegments := strings.Split(path, string(os.PathSeparator))

		dC := drbdConnection{
			Resource:   pathSegments[6],
			RemoteHost: pathSegments[8],
			ResourceID: pathSegments[9],
		}

		procDrbd, err := ioutil.ReadFile(path)
		if err != nil {
			log.Printf("failed to read %s: %s", path, err.Error())
			return nil
		}

		err = parseProcDRBD(procDrbd, &dC)
		if err != nil {
			log.Printf("failed to parse %s: %s", path, err.Error())
			return nil
		}

		connections = append(connections, dC)
		return nil
	})

	return connections
}

var errInvalidOutput = fmt.Errorf("Failed to parse proc_drbd data")

/*
 1: cs:Established ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:119598720 dw:161539164 dr:0 al:2125 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
    resync: used:0/61 hits:25476 misses:922 starving:0 locked:0 changed:592
    act_log: used:0/1237 hits:13283342 misses:15109 starving:0 locked:0 changed:12057
    blocked on activity log: 0/0/0

 0: cs:Established ro:Secondary/Secondary ds:UpToDate/UpToDate C r-----
    ns:0 nr:0 dw:158780236 dr:0 al:1950 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
    resync: used:0/61 hits:2 misses:285 starving:0 locked:0 changed:285
    act_log: used:0/1237 hits:13371654 misses:15913 starving:0 locked:0 changed:11885
    blocked on activity log: 0/0/0

 1: cs:Established ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:119876764 nr:0 dw:154278296 dr:14701504 al:2070 bm:0 lo:0 pe:[0;0] ua:0 ap:[0;0] ep:1 wo:2 oos:0
    resync: used:0/61 hits:25566 misses:719 starving:0 locked:0 changed:485
    act_log: used:0/1237 hits:13318630 misses:41690 starving:0 locked:0 changed:12063
    blocked on activity log: 0/0/0
*/

var bannerRegexp = regexp.MustCompilePOSIX(` [0-9]: cs:([a-zA-Z]+) ro:([^/]+)/([^/]+) ds:([^/]+)/([^/]+) [a-zA-Z] ([\-rs])`)
var kvExtractionRegexp = regexp.MustCompilePOSIX(`(([a-z]+):([0-9]+))+`)

func parseProcDRBD(input []byte, dC *drbdConnection) error {

	if !bannerRegexp.Match(input) {
		return errInvalidOutput
	}

	regexMatches := bannerRegexp.FindAllStringSubmatch(string(input), -1)
	//  1: cs:Established ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
	//[][]string{[]string{" 1: cs:Established ro:Secondary/Primary ds:UpToDate/UpToDate C r", "Established", "Secondary", "Primary", "UpToDate", "UpToDate", "r"}}

	dC.ConnectionStatus = regexMatches[0][1]
	dC.MyRole = regexMatches[0][2]
	dC.RemoteRole = regexMatches[0][3]
	dC.MyDiskStatus = regexMatches[0][4]
	dC.RemoteDiskStatus = regexMatches[0][5]
	if regexMatches[0][6] != "r" {
		dC.Suspended = true
	}

	kvMatches := kvExtractionRegexp.FindAllStringSubmatch(string(input), -1)

	kvs := make([]drbdConnectionKV, 0)
	for _, blob := range kvMatches {
		// blob will look something like:
		// 	[]string{"nr:119598720", "nr:119598720", "nr", "119598720"},
		i, err := strconv.ParseFloat(blob[3], 64)
		if err != nil {
			log.Printf("failed to parse KV number! %#v", blob)
			continue
		}

		kv := drbdConnectionKV{
			Name:  blob[2],
			Value: i,
		}

		kvs = append(kvs, kv)
	}

	dC.KVs = kvs
	return nil
}
